In [18]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 4: High-Entropy Passphrases ---</h2><p>A new system policy has been put in place that requires all accounts to use a <em>passphrase</em> instead of simply a pass<em>word</em>. A passphrase consists of a series of words (lowercase letters) separated by spaces.</p>
<p>To ensure security, a valid passphrase must contain no duplicate words.</p>
<p>For example:</p>
<ul>
<li><code>aa bb cc dd ee</code> is valid.</li>
<li><code>aa bb cc dd aa</code> is not valid - the word <code>aa</code> appears more than once.</li>
<li><code>aa bb cc dd aaa</code> is valid - <code>aa</code> and <code>aaa</code> count as different words.</li>
</ul>
<p>The system's full passphrase list is available as your puzzle input. <em>How many passphrases are valid?</em></p>
</article>


In [19]:
tests = [
    {
        "name": "Example 1",
        "passphrase": "aa bb cc dd ee",
        "expected": True,
    },
    {
        "name": "Example 2",
        "passphrase": "aa bb cc dd aa",
        "expected": False,
    },
    {
        "name": "Example 3",
        "passphrase": "aa bb cc dd aaa",
        "expected": True,
    },
]


def passphrase_valid(passphrase: str) -> bool:
    words_seen = set()
    for word in passphrase.split(" "):
        if word in words_seen:
            return False
        words_seen.add(word)
    return True


run_tests_params(passphrase_valid, tests)

example = """
aa bb cc dd ee
aa bb cc dd aa
aa bb cc dd aaa
"""


def count_valid_passphrases(passphrases: str) -> int:
    return sum(1 for pf in passphrases.strip().splitlines() if passphrase_valid(pf))


assert count_valid_passphrases(example) == 2


Test Example 1 passed, for passphrase_valid.
Test Example 2 passed, for passphrase_valid.
Test Example 3 passed, for passphrase_valid.
Success


In [20]:
with open("../input/day4.txt") as f:
    puzzle = f.read()

print(f"Part I: {count_valid_passphrases(puzzle)}")

Part I: 337


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>337</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>For added security, <span title="Because as everyone knows, the number of rules is proportional to the level of security.">yet another system policy</span> has been put in place.  Now, a valid passphrase must contain no two words that are anagrams of each other - that is, a passphrase is invalid if any word's letters can be rearranged to form any other word in the passphrase.</p>
<p>For example:</p>
<ul>
<li><code>abcde fghij</code> is a valid passphrase.</li>
<li><code>abcde xyz ecdab</code> is not valid - the letters from the third word can be rearranged to form the first word.</li>
<li><code>a ab abc abd abf abj</code> is a valid passphrase, because <em>all</em> letters need to be used when forming another word.</li>
<li><code>iiii oiii ooii oooi oooo</code> is valid.</li>
<li><code>oiii ioii iioi iiio</code> is not valid - any of these words can be rearranged to form any other word.</li>
</ul>
<p>Under this new system policy, <em>how many passphrases are valid?</em></p>
</article>

</main>


In [21]:
tests = [
    {
        "name": "Example 1",
        "passphrase": "abcde fghij",
        "expected": True,
    },
    {
        "name": "Example 2",
        "passphrase": "abcde xyz ecdab",
        "expected": False,
    },
    {
        "name": "Example 3",
        "passphrase": "a ab abc abd abf abj",
        "expected": True,
    },
    {
        "name": "Example 4",
        "passphrase": "iiii oiii ooii oooi oooo",
        "expected": True,
    },
    {
        "name": "Example 5",
        "passphrase": "oiii ioii iioi iiio",
        "expected": False,
    },
]


def passphrase_anagram_valid(passphrase: str) -> bool:
    anagrams_seen = set()
    for word in passphrase.split(" "):
        anagram = "".join(sorted(word))
        if anagram in anagrams_seen:
            return False
        anagrams_seen.add(anagram)
    return True


run_tests_params(passphrase_anagram_valid, tests)

example = """
abcde fghij
abcde xyz ecdab
a ab abc abd abf abj
iiii oiii ooii oooi oooo
oiii ioii iioi iiio
"""


def count_anagram_valid_passphrases(passphrases: str) -> int:
    return sum(
        1 for pf in passphrases.strip().splitlines() if passphrase_anagram_valid(pf)
    )


assert count_anagram_valid_passphrases(example) == 3


Test Example 1 passed, for passphrase_anagram_valid.
Test Example 2 passed, for passphrase_anagram_valid.
Test Example 3 passed, for passphrase_anagram_valid.
Test Example 4 passed, for passphrase_anagram_valid.
Test Example 5 passed, for passphrase_anagram_valid.
Success


In [22]:
print(f"Part II: {count_anagram_valid_passphrases(puzzle)}")

Part II: 231


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>231</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
